File: src/gen_server_call_normal.erl
-module(gen_server_call_normal).
-behaviour(supervisor).
-export([run/3, init/1]).

run(Iters, Timeout, Puts) ->
  {ok, Sup} = supervisor:start_link(?MODULE, []),

  try
    loop(#{}, Iters, Sup, Timeout, Puts)
  after
    gen_server:stop(Sup)
  end.

init(_InitArg) ->
  SupFlags = #{strategy => simple_one_for_one},
  ChildSpecs = [#{id => gen_server_call_normal_server,
                  start => {gen_server_call_normal_server, start_link, []},
                  restart => transient}],
  {ok, {SupFlags, ChildSpecs}}.

loop(Acc, 0, _Sup, _Timeout, _Puts) ->
  Acc;

loop(Acc, Iters, Sup, Timeout, Puts) ->
  {ok, Server} = supervisor:start_child(Sup, [{Iters, Timeout, Puts}]),

  receive
  after
    Timeout -> ok
  end,

  try
    gen_server:call(Server, do_work),
    Acc1 = maps:update_with(success, fun(X) -> X + 1 end, 1, Acc),
    loop(Acc1, Iters - 1, Sup, Timeout, Puts)
  catch
    exit:{Reason, _Request} ->
      Acc2 = maps:update_with(Reason, fun(X) -> X + 1 end, 1, Acc),
      loop(Acc2, Iters - 1, Sup, Timeout, Puts)
  end.

File: lib/gen_server_call_normal.ex
defmodule GenServerCallNormal do
  use DynamicSupervisor

  defmodule Server do
    use GenServer, restart: :transient

    def start_link(arg), do: GenServer.start_link(__MODULE__, arg)

    def init({iteration, timeout, puts?}) do
      Process.send_after(self(), :shutdown, timeout)
      {:ok, {iteration, puts?}}
    end

    def handle_info(:shutdown, {iteration, puts?} = state) do
      if puts?, do: IO.puts("#{iteration}: exiting...")

      {:stop, :normal, state}
    end

    def handle_call(:do_work, _from, state) do
      {:reply, :ok, state}
    end
  end

  @default_opts timeout: 100, iters: 100, puts?: true

  def run(opts \\ []) do
    opts = Keyword.merge(@default_opts, opts)

    timeout = opts[:timeout]
    iters = opts[:iters]
    puts? = opts[:puts?]

    {:ok, supervisor} = DynamicSupervisor.start_link(__MODULE__, [])

    try do
      loop(%{}, iters, supervisor, timeout, puts?)
    after
      DynamicSupervisor.stop(supervisor)
    end
  end

  def init(_init_arg), do: DynamicSupervisor.init(strategy: :one_for_one)

  defp loop(acc, 0, _supervisor, _timeout, _puts?), do: acc

  defp loop(acc, iters, supervisor, timeout, puts?) do
    {:ok, server} = DynamicSupervisor.start_child(supervisor, {Server, {iters, timeout, puts?}})

    Process.sleep(timeout)
    GenServer.call(server, :do_work)

    acc
    |> Map.update(:success, 1, &(&1 + 1))
    |> loop(iters - 1, supervisor, timeout, puts?)
  catch
    :exit, {reason, _request} ->
      acc
      |> Map.update(reason, 1, &(&1 + 1))
      |> loop(iters - 1, supervisor, timeout, puts?)
  end
end
